daemon/k8s_watcher.go

```diff
@@ -1969,6 +1967,43 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
 	}
 }
 
+// missingK8sPodV1 returns all pods considered missing, if the IP doesn't exist in
+// the cache or if there is an endpoint associated with that pod that doesn't
+// contain all labels from the pod.
+func (d *Daemon) missingK8sPodV1(m versioned.Map) versioned.Map {
+	missing := versioned.NewMap()
+	for k, v := range m {
+		pod := v.Data.(*v1.Pod)
+		_, exists := ipcache.IPIdentityCache.LookupByIP(pod.Status.PodIP)
+		// We can only consider that a pods is missing if it doesn't exist
+		// in the identity cache.
+		if !exists {
+			missing.Add(k, v)
+			continue
+		}
+		podNSName := k8sUtils.GetObjNamespaceName(&pod.ObjectMeta)
+
+		podEP := endpointmanager.LookupPodName(podNSName)
+		// Only 1 endpoint in the whole cluster is managing this pod, if it's
+		// not found is due:
+		// - pod is not being managed by this node;
+		// - pod is running on this node but there are no endpoints associated
+		//   with it. This association pod to endpoint association occurs when
+		//   the endpoint is created, by the pkg/workload.
+		if podEP == nil {
+			continue
+		}
+		// If it doesn't contain all pod labels for an endpoint that is managing
+		// that pod then the pod is also considered missing.
+		podLabels := labels.Map2Labels(pod.GetLabels(), labels.LabelSourceK8s)
+		podFilteredLabels, _ := labels.FilterLabels(podLabels)
+		if !podEP.HasLabels(podFilteredLabels) {
```

Comment:
This seems arbitrary. Is the goal to detect when a pod notification contains new information? How would we detect if labels need to be removed?

Reply:
>Is the goal to detect when a pod notification contains new information? 

No, the goal here is: Based on the current source of truth, the kube-apiserver, do we have the correct state in the cilium-agent? This has nothing to do with or without notifications, is to make sure we are correctly in sync with the kube-apiserver.

Reply:
But the `podEp.HasLabels()` is not a correct indication for this as it only checks for the subset and not for an identical match. I may not fully understand the implication of whether we report something as missing or not though.

Reply:
@tgraf >But the podEp.HasLabels() is not a correct indication for this as it only checks for the subset and not for an identical match

We can't check for an identical match because: a) The user might have added more labels to the endpoint, b) There are other labels in the endpoint i.e. namespace labels. What we need to be sure is that all pod labels need to be in the endpoint.

Reply:
What happens if a label is removed from a pod?